### ISSUE TITLE
mask evaluation is really slow

### ISSUE THREAD

**Issue by qianyizhang** [facebookresearch/maskrcnn-benchmark#395]
## 🐛 Bug
prepare_for_coco_segmentation is running extremely slow
![image](https://user-images.githubusercontent.com/12500132/51981620-9a2cc900-24ce-11e9-860b-1ac16960ac74.png)

## To Reproduce
I am testing with given config (e2e_mask_rcnn_X_101_32x8d_FPN_1x.yaml) and adapt to using 4 gpus by following the instructions.

## Expected behavior

while the training and prediction seems fine, the mask evaluation is taking too long, taking ~ 2 hours for 5000imgs.
And the mmlab evaluation of similar setup took about 9 min for inference + 1 min eval box and 1min eval mask.

**Comment by fmassa** [comment]
This is weird.

While I do agree that the mask upsampling + pasting can be improved (see https://github.com/facebookresearch/maskrcnn-benchmark/pull/129 for example) I'd be surprised that it would take that long. Maybe you have way too many detections per class?

**Comment by qianyizhang** [comment]
i am using coco2017_val set.
and the majority of time is consumed by the prepare_for_coco_segmentation function

**Comment by fmassa** [comment]
This step is currently entirely run on the CPU and not parallelized, and can definitely be improved.

Are there any other changes that you did to the implementation? Also, did you change the value of
https://github.com/facebookresearch/maskrcnn-benchmark/blob/ebc593a2985c492b9a2a32486cc839f954a03cfa/maskrcnn_benchmark/config/defaults.py#L196
?

**Comment by qianyizhang** [comment]
no, the only thing I changed is fixing the batchsize and learning rate accordingly to accommodate 4 gpu training.
On the other hand, the score is on par with the reported value in the MODEL_ZOO.md, so nothing is wrong I guess.
Can you run a eval on your side, to confirm about the time?

PS: I noticed pretty much all cpus are occupied during eval (24 cores server)

**Comment by fmassa** [comment]
One thing I realized is that maybe https://github.com/facebookresearch/maskrcnn-benchmark/pull/180 has introduced some serious regressions by replacing PIL with a pytorch function (interpolate).
Can you check if reverting this change makes things faster for you?

**Comment by qianyizhang** [comment]
the newly introduced SegmentationMask speed up evaluation x20, you can close this issue.

**Comment by lzrobots** [comment]
@qianyizhang Hi how did you solve this problem?

**Comment by andyhahaha** [comment]
@qianyizhang @fmassa  it takes me 4 hours to evaluate the mask results.
And most of time is consumed by the prepare_for_coco_segmentation function too.
Is there any speed up solution?
Such as multiprocessing?

**Comment by hkzhang95** [comment]
> @qianyizhang @fmassa it takes me 4 hours to evaluate the mask results.
> And most of time is consumed by the prepare_for_coco_segmentation function too.
> Is there any speed up solution?
> Such as multiprocessing?

I find that [paste_mask_in_image](https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/maskrcnn_benchmark/modeling/roi_heads/mask_head/inference.py#L176) is slow, especially the initialization of [im_mask](https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/maskrcnn_benchmark/modeling/roi_heads/mask_head/inference.py#L150).

Then I try to initialize a big `im_mask` and paste the masks in corresponding images as follows:

```
        if len(masks) > 0:
            res = torch.zeros(
                (len(masks), 1, im_h, im_w), dtype=torch.uint8,
                device=masks.device)
            for idx, (m, b) in enumerate(zip(masks, boxes)):
                paste_mask_in_image(
                    res[idx, 0, :, :], m[0], b, im_h, im_w, self.threshold)
```

And I find it speeds up by about 10x (previous speed is 1 it/s). Maybe you can have a try.
